docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,150 @@
+    springboot dev-tools.
+
+2. If this is a redesign, what are the drawbacks of the current implementation?
+  - This is not a redesign, but a new sync mode that may or may not be available
+    in the skaffold.yaml directly to users. It could simply be an internal API
+    that is usuable by builders like Jib.
+  - This is similar to `_smart_` described in [sync-improvements](sync-improvements.md)
+
+3. Is there any another workaround, and if so, what are its drawbacks?
+  - Currently one can create a docker file that only copies specific build
+    results into the container and relies on a local build to generate those
+    intermediate artifacts. This requires a user to trigger a first build
+    manually before docker kicks in to containerize the application. While
+    it may be possible to automate this (for example: gradle --continuous), it
+    is not an acceptable solution to require manual external processes for
+    a build to succeed.
+
+4. Mention related issues, if there are any.
+  - This is not trying to solve the problem of dealing with a multistage
+    dockerbuild. Intermediate build artifacts might still be possible to
+    determine, however that would require an extra mechanism to do a partial
+    docker build and sync files from a built container to a running container --
+    something we do not intend to cover here.
+
+#### Problems with current API/config
+The current `sync` system has the following problems:
+1. No way to trigger local external processes - for example, in jib, to build a
+   container, one would run `./gradlew jib`, but to update class files so the
+   system may sync them, one would only be required to run `./gradlew classes`.
+2. No way to tell the system to sync non build inputs. Skaffold is
+   normally only watching `.java` files for a build, in the sync case, we want
+   it to watch `.java` files, trigger a partial build, and sync `.class` files
+   so a remote server can pick up and reload the changes.
+
+## Design
+
+#### Hack it
+To get close to the functionality we want, without modifying skaffold at all, a
+Dockerfile which depends on java build outputs could be used, like:
+```
+FROM openjdk:8
+COPY build/dependencies/ /app/dependencies
+COPY build/classes/java/main/ /app/classes
+
+CMD ["java", "-cp", "/app/classes:/app/dependencies/*", "hello.Application"]
+```
+
+with a skaffold sync block that looks like:
+```
+sync:
+  manual:
+    - src: "build/classes/java/main/**/*.class"
+      dest: "/app/classes"
+      strip: "build/classes/java/main/"
+```
+
+A user's devloop then looks like this:
+
+1. run `./gradlew classes copyDependencies`
+2. run `skaffold dev`
+3. *make changes to some java file*
+4. run `./gradlew classes`
+5. *skaffold syncs files*
+
+which is far from ideal.
+
+#### Fix it?
+
+What we might want, is an API that can accept configuration that looks like:
+
+```
+sync: auto {
+  indirect:
+  - command: "./gradlew jib classes"
```

Comment:
From a usability perspective, it is still quite some intellectual overhead to specify the build commands in the right way and define the proper sync rules. My gut feeling is that this procedure is a lot different from what most developers are used to, especially for java (if you do SpringBoot, you never care about `.class` files). Do you plan a further simplification so that the sync works out of the box for, say, SpringBoot?

Reply:
I think this config is mostly a manifestation of what the API would be for a builder like jib to define all this - not necessarily a user. So `jib` would just tell the sync system what the config is, and it would go about doing what needs to be done.

I'm not opposed to it, but I agree, it's fairly involved for a user to configure this.

Reply:
I'm still confused. Will the user be able to configure all these in `skaffold.yaml`, or is it that this doc just explains a proposed internal sync API in the form? In either case, what if the user configures other current sync options in their `skaffold.yaml`? Will they conflict?

Reply:
In much the same vein, I'd recommend to list a few other use cases for the proposed yaml changes beside better Jib support. It sounds as if your main goal is to achieve better Jib integration, but that could be done with less user-facing changes. I'm a bit worried that the current proposal is very powerful, but also rather complex (both for users and maintenance).
If the sole intended use for this is Jib, there could be a simpler option without so many user-facing changes.

Reply:
sure, I just needed a way to describe the API in a way that could be read. There may not be any need at all to expose this beyond an internal API. It also helps explain how this might be extended beyond just jib usage (though that might be the only usecase for now).